### DISCUSSION TITLE
Fetch data from old file

### DISCUSSION THREAD

**Discussion by nirbar89** [simolus3/drift#2531] [Q&A] [answered]
Hello,
I have an application which running for few months and now I'm starting to create a new version with new features (almost new app)
For that I'm going to use a clean database which I'm going to create a new file.

BUT I need to get a value from a specific column in a specific table from my old db file. 

what would be the best way to do that?

**Comment by simolus3** [comment]
If it's just a simple fetch, I'd probably recommend reading the old values with the plain `sqlite3` package: You can open the old database with `sqlite3.open(oldDatabaseFile)` and then use `select()` on the database returned by `open()`.

You _could_ use a second drift database class without a migration copying the table definition for the table as well. Then you'd open an instance of that database class, read values, map them into the classes generated for the new database, and save them. This is probably safer than reading them manually, but also requires you to keep a lot of existing database code from the old version around.

**Comment by nirbar89** [comment]
Ok, i choose the first option and use sqlite3.

My question now is where to use it...

i can do it in onCreate like this:

```
class AppDatabase extends _$AppDatabase {
  AppDatabase() : super(_openConnection());

  static Future<AppDatabase> init() async {
    final database = AppDatabase();
    await database.doWhenOpened((e) => null);
    return database;
  }

  @override
  int get schemaVersion => 1;

  @override
  MigrationStrategy get migration {
    return MigrationStrategy(
      onCreate: (Migrator m) async {
        await m.createAll();
        final dbFolder = await getApplicationDocumentsDirectory();
        final file = File(p.join(dbFolder.path, 'OLD_db.sqlite'));
        if (file.existsSync()) {
          try {
            final db = sqlite3.open(file.path);
            final oldVal = db.select("SELECT id, val FROM old_table_name WHERE id=:p1 LIMIT 1", [0]);
            await into(newModelTable).insert(Model(id: 0, val: oldVal.first.values[1] as String?));
            db.dispose();
            await file.delete();
          } catch (e, s) {
          }
        }
      },
      onUpgrade: (Migrator m, int from, int to) async {
      },
      beforeOpen: (details) async {
        await customStatement('PRAGMA foreign_keys = ON');
        await customStatement('PRAGMA journal_mode = WAL;');
      },
    );
  }

}

LazyDatabase _openConnection() {
  // the LazyDatabase util lets us find the right location for the file async.
  return LazyDatabase(() async {
    // put the database file, called db.sqlite here, into the documents folder
    // for your app.
    final dbFolder = await getApplicationDocumentsDirectory();
    final file = File(p.join(dbFolder.path, 'NEW_db.sqlite'));
    return NativeDatabase.createInBackground(file);
  });
}

```
OR: do the same in beforeOpen and use details.wasCreated as a condition to execute the same method

Does anyone know which is best practice?

**Reply by simolus3** [reply, marked as answer]
> OR: do the same in beforeOpen and use details.wasCreated as a condition to execute the same method

That will pretty much work out to the same thing. I think it kind of belongs in `onCreate` because that highlights that this is put in a place that should only run once. `beforeOpen` is something that runs multiple times.

**Reply by nirbar89** [reply]
That's what I thought.

Thank you.
